torch/distributed/_composable/fsdp/_fsdp_collectives.py

```diff
@@ -167,17 +167,19 @@ def foreach_reduce_scatter(
         for padded_unsharded_size, fsdp_param in zip(
             padded_unsharded_sizes, fsdp_params
         ):
-            padded_sharded_numel = padded_unsharded_size.numel() // world_size
-            sharded_numel = fsdp_param.sharded_size.numel()
-            new_sharded_grad = reduce_scatter_output[
-                flat_grad_offset : flat_grad_offset + sharded_numel
-            ].view(fsdp_param.sharded_size)
+            new_sharded_grad = torch.as_strided(
+                reduce_scatter_output,
+                size=fsdp_param.sharded_size,
+                stride=fsdp_param.sharded_stride,
```

Comment:
This is very bizarre. I isolated that this change from
```
sharded_numel = fsdp_param.sharded_size.numel()
new_sharded_grad = reduce_scatter_output[
    flat_grad_offset : flat_grad_offset + sharded_numel
].view(fsdp_param.sharded_size)
```
to
```
new_sharded_grad = torch.as_strided(
    reduce_scatter_output,
    size=fsdp_param.sharded_size,
    stride=fsdp_param.sharded_stride,
    storage_offset=flat_grad_offset,
)
```
as affecting correctness for a 2D unit test downstream. The strange thing is that I checked that the two versions pass `torch.equal()` and have the same size/stride/data pointer, so I am not sure where the difference is coming from.

cc: @yifuwang

Reply:
I have further found that (1) it seems to only affect the 2D-parallel case and (2) it seems to only affect uneven case (when dim-0 is not divisible by world size).

Reply:
It seems possible that the `DTensor`'s `_local_tensor` stride is not correct for uneven case 🤔 

I am seeing some weirdness that needs to be investigated more. Like the uneven chunk from `torch.chunk()` shows stride `(5, 1)` despite shape `(1, 4)` (so we expect stride `(4, 1)`).

Reply:
I figured it out: the `fsdp_param.sharded_stride` was being taken from the `sharded_param` chunked from the `DTensor._local_tensor`, which had the incorrect stride in the uneven sharding case. Therefore, the `fsdp_param.sharded_stride` being passed to `torch.as_strided` was wrong. I am not sure why the direct `.strides()` check did not fail when comparing the two approaches -- I must have done something wrong.

I filed an issue and added a workaround in this PR.

Reply:
>  figured it out: the fsdp_param.sharded_stride was being taken from the sharded_param chunked from the DTensor._local_tensor

Oh yea I noticed that when I first saw it. The PR didn't have usage for it yet so I wasn't sure if it was intended that way.